qa/tasks/caasp.py

```diff
@@ -0,0 +1,137 @@
+        log.debug("beginning of constructor method")
+        self.ctx['roles'] = self.ctx.config['roles']
+        self.log = log
+        self.remotes = self.cluster.remotes
+        self.mgmt_remote = get_remote_for_role(self.ctx, "skuba_mgmt_host.0")
+        self.ssh_priv = 'tempkey.rsa'
+        self.ssh_pub = 'tempkey.rsa.pub'
+        self.set_agent = "eval `ssh-agent` && ssh-add ~/.ssh/%s && " % self.ssh_priv
+
+    def __ssh_setup(self):
+        log.debug("Executing SSH setup")
+        self.__ssh_gen_key()
+        self.__ssh_copy_priv()
+        self.__ssh_copy_pub_to_caasp()
+
+    def __ssh_gen_key(self):
+        if os.path.isfile(self.ssh_priv):
+            os.remove(self.ssh_priv)
+        if os.path.isfile(self.ssh_pub):
+            os.remove(self.ssh_pub)
+        os.system('ssh-keygen -t rsa -b 2048 -P "" -f %s' % self.ssh_priv)
+#            subprocess.check_output(['ssh-keygen', '-t', 'rsa', '-b', '2048', '-f', self.ssh_priv])
+#            res = subprocess.check_output(["ls"])
+#            log.debug(res)
+#        except subprocess.CalledProcessError as e:
+#            log.debug('error generating key')
+#            log.debug(e.output)
+
+    def __ssh_copy_priv(self):
+        log.debug("Executing __ssh_copy_priv")
+        try:
+            subprocess.check_output(['scp', self.ssh_priv, '%s:.ssh/' %
+                                     self.mgmt_remote])
+        except subprocess.CalledProcessError as e:
+            log.debug(e.output)
+
+    def __ssh_copy_pub(self, remote):
+        log.debug("Executing __ssh_copy_pub to %s" % remote.hostname)
+        cmd = "ssh-copy-id -i %s %s" % (self.ssh_pub, remote)
+        log.debug(cmd)
+        os.system(cmd)
+ #        try:
+ #            subprocess.check_output(
+ #                ['ssh-copy-id', '-i', self.ssh_pub, '%s' % remote])
+ #        except subprocess.CalledProcessError as e:
+ #            log.debug(e.output)
+
+    def __ssh_copy_pub_to_caasp(self):
+        for i in range(sum(1 for x in all_roles_of_type(
+                self.ctx.cluster, 'caasp_master'))):
+            remote = get_remote_for_role(
+                self.ctx, "caasp_master." + str(i))
+            self.__ssh_copy_pub(remote)
+        for i in range(sum(1 for x in all_roles_of_type(
+                self.ctx.cluster, 'caasp_worker'))):
+            remote = get_remote_for_role(
+                self.ctx, "caasp_worker." + str(i))
+            self.__ssh_copy_pub(remote)
+
+    def __create_cluster(self):
+        master_remote = get_remote_for_role(self.ctx, "caasp_master.0")
+        commands = [
+            "ssh-add -L",
+            "skuba cluster init --control-plane {} cluster".format(master_remote.hostname),
+            "cd cluster && skuba node bootstrap --user ubuntu --sudo --target {} my-master".format(
+                master_remote.hostname),
+        ]
+        for command in commands:
+            self.mgmt_remote.sh("%s %s" % (self.set_agent, command))
```

Comment:
Is there a need to run commands sequentially in such a way? Why not run everything as single script, like?
```
self.mgmt_remote.sh("
            {agent} ssh-add -L
            {agent} skuba cluster init --control-plane {master} cluster
            {agent} cd cluster && skuba node bootstrap --user ubuntu --sudo --target {master} my-master
            ".format(agent=self.set_agent, master=master_remote.hostname)
```
This code is a bit easier to understand, however, I would get rid of this multiple agent prerun stuff. Once you've set the keypairs correctly, do you need it at all?